### PR TITLE
fix for continous lines in macros

### PR DESCRIPTION
**Pull request by M0ses**

### CODE REVIEW

**Comment by mlschroe** [comment]
Isn't that a bit too special-coded for that %define %if case? It may be better to simply ignore all %define statements (including their continuation lines).

**Comment by M0ses** [comment]
@mlschroe better ?

**Comment by mlschroe** [comment]
Dunno, seems way to complicated ;)

**Comment by marcus-h** [comment]
On 2018-01-17 12:50:28 +0000, Frank Schreiner wrote:
> @M0ses requested your review on: openSUSE/obs-service-source_validator#68 fix for continous lines in macros.
> 
Hmm maybe we shouldn't scatter the line continuation handling and
handle it in one place. The current code "fails" for ill-defined
macros like

%define ifelsepart \
   %if 0 \
   x \
   %else
%define endifpart echo "xxx" \
   %endif

%ifelsepart
%endifpart

Something like this should do the trick:

$start_cont = 1 if /^\s*%(?:define|global)\s+.*\\$/;
if (/\\$/) {
  push @$spec, $_;
  next;
} elsif ($start_cont) {
  push @$spec, $_;
  $start_cont = 0;
  next;
}

The attached patch is based on your commit 1e7ce46.

**Comment by marcus-h** [comment]
On 2018-01-17 22:39:14 +0100, Marcus Hüwe wrote:
> The attached patch is based on your commit 1e7ce46.

Hrm... since github does not like attachments, here's
the patch:


diff --git a/helpers/spec_query b/helpers/spec_query
index bfb2efc..64ee389 100755
--- a/helpers/spec_query
+++ b/helpers/spec_query
@@ -27,14 +27,16 @@ sub prepare_spec {
   my @ifs;
   my $start_cont = 0;
   while (<F>) {
-    if ($start_cont && /\\$/) {
+    $start_cont = 1 if /^\s*%(?:define|global)\s+.*\\$/;
+    if (/\\$/) {
       push @$spec, $_;
       next;
-    } elsif (/^\s*%define/) {
-      $start_cont = 1 if (/\\$/);
+    } elsif ($start_cont) {
       push @$spec, $_;
+      $start_cont = 0;
       next;
-    } elsif (/^%if\s+.*%\{?name/ && $keep_name_conditionals) {
+    }
+    if (/^%if\s+.*%\{?name/ && $keep_name_conditionals) {
       unshift @ifs, $st_ifname;
     } elsif (/^%if/) {
       unshift @ifs, $st_if;
@@ -47,12 +49,6 @@ sub prepare_spec {
       }
       next;
     } elsif (/^\s*%endif/) {
-      if ( $start_cont) {
-        $start_cont=0;
-        # keep %endif if it ends a line continuation
-        push @$spec, $_;
-        next;
-      }
       die("got %endif without %if\n") unless @ifs;
       push @$spec, "%endif" if $ifs[0] == $st_ifname;
       shift @ifs;
@@ -70,7 +66,6 @@ sub prepare_spec {
       $seen{$tag} = 1;
     }
     push @$spec, $_;
-    $start_cont = 0 unless (/\\$/);
   }
   die("got %if without %endif\n") if @ifs;
   close(F) || die("close: $!\n");

**Comment by mlschroe** [comment]
How about simply

```
 while (<F>) {
  if (/^\s*%(?:define|global)) {
    $_ = <F> while /\\$/;
    next;
  }
}
```

And no weird start_cond handling at all...

**Comment by M0ses** [comment]
@marcus-h 

Thanks for your suggestions, but I followed @mlschroe 's advice and hope you will like it too.

Two questions are open (ATM)

1. I decativated the warnings for Build::Rpm::parse (https://github.com/openSUSE/obs-service-source_validator/pull/68/files#diff-a29cff436429d6f0ae0849a302a4b76aL95) 
  Is this ok for you?
2. Why is the `chomp` done so late (https://github.com/openSUSE/obs-service-source_validator/pull/68/files#diff-a29cff436429d6f0ae0849a302a4b76aR60)
  Is there any specific reason for that?

I hope, we could increase the test coverage soon, because it`s very hard ATM for new developers (as I see myself) to get an idea, which goals should be achieved

File: helpers/spec_query [hunk 1]
```diff
@@ -25,35 +25,38 @@ sub prepare_spec {
...
-        next;
+      next;
     } elsif (/^\s*%else/) {
-      $start_cont = 0 unless (/\\$/);
       die("got %else without %if\n") unless @ifs;
       if ($ifs[0] == $st_ifname) {
-        push @$spec, "%endif";
-        $ifs[0] = $st_if;
+	push @$spec, "%endif";
+	$ifs[0] = $st_if;
```
Comment (marcus-h) [inline helpers/spec_query]: Are the tabs intended?

**Comment by marcus-h** [comment]
I just enabled the warnings so that we are aware of potential issues. Is there a specific reason why you
want to disable them? (It's ok for me if you disable them...) You can also move the chomp to the top.
In general, the patch looks ok to me.

File: helpers/spec_query (hunk 1 above)
Comment (M0ses) [reply helpers/spec_query]: No, they were invoked by my general vim settings (which we are also using e.g. in open-build-service). I will fix that.

**Comment by M0ses** [comment]
I would like to disable them, because the input for Build::Rpm::parse is not the original file.

In the test cases I found files which pass Build::Rpm::parse without a problem, but because of the filtering done by prepare_parse the warnings occurred. Using line numbers which have nothing to do with the original input file and giving no additional information about the context. In the actual state this is more confusing then helpful. 

Thats why I want to deactivate them.

**Pull request merged**